doc: CLI spec for trust policy management (DO NOT REVIEW) (#505)
notaryproject/notation: specs/commandline/policy.md

            "trustedIdentities": [                                      // Identities that are trusted to sign the artifact.
                "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
            ]
        },
        {
            "name": "wabbit-networks-prod",                             // Name of the 2nd policy.
            "registryScopes": [ "prod.wabbitnetworks.io/net-monitor" ],       
            "signatureVerification": {                                
                "level": "permissive"
            },
            "trustStores": [ "ca:wabbit-networks-prod" ],                  
            "trustedIdentities": [                                    
                "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Prod, CN=wabbit-networks.io"
            ]
        }
    ]
}
```

## Outline

### notation policy command

```text
Manage trust policies for signature verification.

Usage:
  notation policy [command]

Available Commands:
  add           Add trust policies.
  delete        Delete trust policies.
  list          List trust policies.

>   update        Update trust policies

Author: I think maybe it's better to remove the periods.

Flags:
  -h, --help   help for policy
```

### notation policy add

```text
Add trust policies.

Usage:

>   notation policy add [flags] <policy_name>

Reviewer: If we do ` notation policy add json_file_path`, and there are more than one trust policy statements in the file, they will have different policy names right? It seems that the current design is only taking one <policy_name> as argument. What should we do in this case? According to our original design https://hackmd.io/AP9Xz9AbRpOy6ASpbok23A#Manage-Trust-Policy, we could have something like:
```
Usage:
    notation policy add (--input <policy_file_path> | --name <policy_name>) [flags]
Flags:
   --input string  path to a trust policy file in JSON format
   --name  string  name of the trust policy statement
   ...
```

Flags:
      --cert-file          string   optional certificate filepath
      --custom-level       string   optional custom level to based existing verification level, e.g. "authenticity=log,expiry=log"
  -h, --help                        help for add
      --id                 string   optional trust identity aka certificate subject info
      --repo               string   optional repository URI (default "*")

>       --trust-store        string   required trust store in format "<trust_store_type>:<trust_store_name>", e.g. "ca:my_trust_store"

Reviewer: Same here. It's an array.
```suggestion
      --trust-store        stringArray   required trust stores in format <trust_store_type>:<trust_store_name>
```
examples can be moved into the `Long` message of the command.

>       --verification-level string   optional verification level, options: "strict", "permissive", "audit", "skip" (default "strict")

Reviewer: ```suggestion
      --verification-level string   verification level, options: "strict", "permissive", "audit", "skip" (default "strict")
```

```

### notation policy delete

```text
Delete trust policies. User cannot delete all the trust policies, at least one trust policy should be configured for signature verification.

Usage:
  notation policy delete [flags] <policy_name>...

Flags:
  -h, --help   help for delete
```

### notation policy list

```text
List trust policies by names

Usage:
  notation policy list [flags]

Aliases:
  list, ls

Flags:
      --cert-file   string       optional certificate filepath
      --details                  optional list the details of trust policies
  -h, --help                     help for list
      --id          string       optional trust identity aka certificate subject info
      --name        string       optional trust policy name
      --reference   string       optional reference to the artifact
      --repo        string       optional repository URI, e.g. localhost:5000/namespace/repo_name
      --trust-store string       optional trust store in format "<trust_store_type>:<trust_store_name>", e.g. "ca:my_trust_store"
```

### notation policy update

```text
Update the existing trust policies.

Usage:
  notation policy update [flags] <policy_name>

Flags:

>       --cert-file          string   optional certificate filepath
>       --custom-level       string   optional custom level to based existing verification level, e.g. "authenticity=log,expiry=log"
>   -h, --help                        help for add
>       --id                 string   optional trust identity aka certificate subject info
>       --repo               string   optional repository URI
>       --trust-store        string   optional trust store in format "<trust_store_type>:<trust_store_name>", e.g. "ca:my_trust_store"
>       --verification-level string   optional verification level, options: strict, permissive, audit, skip

Reviewer: ```suggestion
     --custom-level       stringArray   {key}={value} pairs that represent a custom level to override existing verification level
-h,  --help                             help for add
     --id                 stringArray   trust identities, user trusted signing certificate subjects
     --scope              stringArray   repository URIs to which the policy applies
     --trust-store        stringArray   trust stores in format <trust_store_type>:<trust_store_name>
     --verification-level string        verification level, options: "strict", "permissive", "audit", "skip"
```
IMO, `notation policy` command's logic should be separated from `notation cert`. The `--cert-file` flag in `notation policy` command might confuse users between these two commands.

```

## Usage

### Add a trust policy

To add a trust policy, users need to specify the **policy name** as mandatory argument and configure other properties accordingly. The following table shows how to configure properties and default values.

| Property Name       | Necessity | Default value | flag to configure               |
| ------------------- | --------- | ------------- | ------------------------------- |
| `trustStores`       | Required  | N/A           | `--trust-store`                 |

> | `registryScopes`    | Optional  | `"*"`         | `--repo`                        |

Reviewer: ```suggestion
| `registryScopes`    | Optional  | `["*"]`         | `--repo`                        |
```

| `level`             | Optional  | `"strict"`    | `--verification-level`          |
| `override`          | Optional  | nil           | `--custom-level`                |

> | `trustedIdentities` | Optional  | `"*"`         | `--id` or `--cert-file` or both |

Reviewer: ```suggestion
| `trustedIdentities` | Optional  | `["*"]`         | `--id` or `--cert-file` or both |
```

```shell  
notation policy add --repo "dev.wabbitnetworks.io/net-monitor" --trust-store "ca:wabbit-network-dev" --id "C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io" --verification-level "strict" wabbit-network-dev
```

The execution of `add` fails in one of below cases:

- The policy name exists.

> - There can only be one trust policy that uses a global scope, that is, the value of `registryScopes` is `*`.

Reviewer: ```suggestion
- There can only be one trust policy that uses a global scope, that is, the value of `registryScopes` is `[*]`.
```
same applies to other places below.

Upon successful execution, the added trust policy is printed out. For example:

In text format

```text
"name": "wabbit-networks-dev"
"registryScopes": "dev.wabbitnetworks.io/net-monitor"
"level": "strict"
"trustStores": "ca:wabbit-networks-dev"
"trustedIdentities": "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
```

In json format

```json
{
  "name": "wabbit-networks-dev",                              
  "registryScopes": [ "dev.wabbitnetworks.io/net-monitor" ],                                  
  "signatureVerification": {                                  
      "level": "strict"
  },
  "trustStores": [ "ca:wabbit-networks-dev" ],                
  "trustedIdentities": [                                      
      "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
  ]
}
```

> ### Add a trust policy with default values

Reviewer: It seems that we are missing the scenario `notation policy add --policy-file policy_file_path`?

In text format

```text
"name": "wabbit-networks-dev"
"registryScopes": "*"
"level": "strict"
"trustStores": "ca:wabbit-networks-dev"
"trustedIdentities": "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
```

In json format

```json
{
  "name": "wabbit-networks-dev",                              
  "registryScopes": [ "*" ],                                  
  "signatureVerification": {                                  
      "level": "strict"
  },
  "trustStores": [ "ca:wabbit-networks-dev" ],                
  "trustedIdentities": [                                      
      "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
  ]
}
```

### Add a trust policy with custom verification level

Users can override the default verification level by using flag `--custom-level`.

```shell
# customize the verification level based on the default verification level `strict`
notation policy add --trust-store "ca:wabbit-network-dev" --custom-level "expiry=log,authenticity=log" wabbit-network-dev
```

Upon successful execution, the added trust policy is printed out. For example:

In text format

```text

> "name": "wabbit-networks-dev"

Reviewer 2: ```suggestion
Name: "wabbit-networks-dev"
```
Is it more clear?

"registryScopes": "*"
"level": "strict"
"override": "expiry=log", "authenticity=log"
"trustStores": "ca:wabbit-networks-dev"

> "trustedIdentities": "*"

Reviewer 2: How to represent an array of string?

    },
    "trustStores": [ "ca:wabbit-networks-prod" ],                  
    "trustedIdentities": [                                    
        "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Prod, CN=wabbit-networks.io"
    ]
}
```

### List trust policies with specified certificate used

```shell
notation policy list --cert-file "./wabbit-networks-prod.crt"
```

An example of output messages:

In text format:

```text
"name": "wabbit-networks-prod"
"registryScopes": "prod.wabbitnetworks.io/net-monitor"
"level": "permissive"
"trustStores": "ca:wabbit-networks-prod"
"trustedIdentities": "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Prod, CN=wabbit-networks.io"
```

In JSON format:

```json
{
    "name": "wabbit-networks-prod",
    "registryScopes": [ "prod.wabbitnetworks.io/net-monitor" ],       
    "signatureVerification": {                                
        "level": "permissive"
    },
    "trustStores": [ "ca:wabbit-networks-prod" ],                  
    "trustedIdentities": [                                    
        "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Prod, CN=wabbit-networks.io"
    ]
}
```

### Delete trust policies

> Users cannot delete all the trust policies, at least one trust policy should be kept for signature verification. Deletion SHOULD fail if users intend to delete all the trust policies.

Reviewer: ```suggestion
Users cannot delete all the trust policies, at least one trust policy should be kept for signature verification. Deletion MUST fail if users intend to delete all the trust policies.
```

```shell
# Delete one trust policy
notation policy delete wabbit-network-dev

# Delete multiple trust policies

> notation policy delete wabbit-network-dev wabbit-network-prod

Reviewer 2: What is the output of delete command?

                     "authenticity": "log"
                }
            },
            "trustStores": [ "ca:wabbit-networks-dev" ],                // The trust stores that contains the X.509 certificates
            "trustedIdentities": [                                      // Identities that are trusted to sign the artifact.
                "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
            ]
        },
        {
            "name": "wabbit-networks-prod",                             // Name of the second policy statement.
            "registryScopes": [ "prod.wabbitnetworks.io/net-monitor" ],       
            "signatureVerification": {                                
                "level": "permissive"
            },
            "trustStores": [ "ca:wabbit-networks-prod" ],                  
            "trustedIdentities": [                                    
                "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Prod, CN=wabbit-networks.io"
            ]
        }
    ]
}
```

## Outline

### notation policy command

```text
Manage trust policies for signature verification.

Usage:
  notation policy [command]

> Available Commands:

Reviewer 3: We also need a mechanism to export this trust policy to a file so that users can share/deploy them later in the systems like ratify etc.

This can be achieved by either displaying trust policy location or adding an `export` command or adding a flag `--export` in the list command,

Author: @priteshbandi Maybe we can reuse `notation policy show`, currently it is called `notation policy list`, maybe `show` is better, so that, by default, the content of trust policy file is displayed. Does it make sense?

  add           Add trust policies
  delete        Delete trust policies
  list          List trust policies
  update        Update trust policies

Flags:
  -h, --help   help for policy
```

### notation policy add

```text
Add trust policies.

Usage:
  notation policy add (--file <policy_file_path> | --name <policy_name>) [flags]

Flags:

>       --custom-level       stringArray   {key}={value} pairs that represent a custom level to override existing verification level

Reviewer 3: For all the fields that have type `stringArray` can we use `string`? customer can use them multiple times to pass multiple values. Also, in this case default value for --scope becomes `*`

  -f  --file               string        path to a trust policy file in JSON format
  -n  --name               string        name of the trust policy statement
  -h, --help                             help for add
      --scope              stringArray   repository URIs to which the policy applies (default ["*"])
      --trust-store        stringArray   trust stores in format "<trust_store_type>:<trust_store_name>"
      --verification-level string        verification level, options: "strict", "permissive", "audit", "skip" (default "strict")

>       --x509-cert          stringArray   paths to x509 certificate file that certificate subject is retrieved from

Reviewer 3: How are expecting user to use use `--x509-cert`? IMO software publisher will distribute software and share the identity(not the certificate) that consumer needs to trust.

Author: @priteshbandi The `--x509-cert` here is designed for one scenario: Users have the certificate file, not the identity, then with this flag, `notation policy` command can retrieve the identity from the certificate file for the users. One question for your case: If it is the identity shared to users, how can users add certificate file using `notation cert add` command?

>       --x509-id            stringArray   trust identities, user trusted x509 certificate subjects

Reviewer 3: should we rename this to `--trusted-identity`

```

### notation policy delete

```text

> Delete trust policies. User cannot delete all the trust policies, at least one trust policy should be configured for signature verification.

Reviewer 3: `user cannot delete all the trust policies, at least one trust policy should be configured for signature verification.`
Instead of restricting it should add a disclaimer that for signature verification, notation needs at least one trust-policy?

Usage:
  notation policy delete [flags] <policy_name>...

Flags:
  -h, --help   help for delete
```

### notation policy list

```text
List trust policies by names

Usage:
  notation policy list [flags]

Aliases:
  list, ls

Flags:

>       --details                 list the details of trust policies

Reviewer 4: We could have something like `--raw` or `--export` to print the `trustpolicy.json` file directly.

  -h, --help                    help for list
  -n, --name        string      name of the trust policy statement

>       --reference   string      reference to the artifact

Reviewer 3: Why do we need this flag?

      --scope       string      repository URI, e.g. localhost:5000/namespace/repo_name
```

### notation policy update

```text
Update the existing trust policies.

Usage:
  notation policy update (--file <policy_file_path> | --name <policy_name>) [flags]

Flags:
      --custom-level       stringArray   {key}={value} pairs that represent a custom level to override existing verification level
  -f  --file               string        path to a trust policy file in JSON format
  -n  --name               string        name of the trust policy statement
  -h, --help               help for add
      --scope              stringArray   repository URIs to which the policy applies (default ["*"])
      --trust-store        stringArray   trust stores in format "<trust_store_type>:<trust_store_name>"
      --verification-level string        verification level, options: "strict", "permissive", "audit", "skip" (default "strict")

>       --x509-cert          stringArray   paths to x509 certificate file that certificate subject is retrieved from

Reviewer 3: ditto

      --x509-id            stringArray   trust identities, user trusted x509 certificate subjects
```

## Usage

### Add a trust policy by configuring the properties from command line

The following table shows how to configure properties for a trust policy and default values.

> | Property Name       | Necessity | flag to configure                    | Default value |

Reviewer 3: ```suggestion
| Property Name       | Required | flag to configure                    | Default value |
```

use yes and no for each propery

| ------------------- | --------- | ------------------------------------ | ------------- |
| `name`              | Required  | `--name`                             | N/A           |
| `trustStores`       | Required  | `--trust-store`                      | N/A           |
| `registryScopes`    | Optional  | `--scope`                            | `["*"]`       |
| `level`             | Optional  | `--verification-level`               | `"strict"`    |
| `override`          | Optional  | `--custom-level`                     | nil           |
| `trustedIdentities` | Optional  | `--x509-id` or `--x509-cert` or both | `["*"]`       |

```shell  
notation policy add --name wabbit-network-dev --scope "dev.wabbitnetworks.io/net-monitor" --trust-store "ca:wabbit-network-dev" --x509-id "C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io" --verification-level "strict"
```

The execution of `add` fails in any of below cases:

> - The trust policy name exists.

Reviewer 3: ```suggestion
- The trust policy name is not specified.
```

- More than one trust policy that uses a global scope, that is, the value of `registryScopes` is `["*"]`.
- The values of `--x509-id` or `--x509cert` overlap. For example, the following two identity values are overlapping:
  - "C=US, ST=WA, O=wabbit-network.io, OU=org1"
  - "C=US, ST=WA, O=wabbit-network.io"

Upon successful execution, the added trust policy is printed out. For example:

> In text format

Reviewer 4: Do we really need text format for the `notation policy` command set? To be honest, the text format is less readable than the json format for a human.

  ]
}
```

Users can also use `notation policy list` to confirm the trust policies are added.

### Add a trust policy with default values

Based on the above table, users can leveraging default values when adding a trust policy. Note that the default value of  `registrySopes` is `["*"]` called global scope, which means this policy applies to all the artifacts. There can only be one trust policy that uses a global scope.

```shell
notation policy add --name wabbit-network-dev --trust-store "ca:wabbit-network-dev"
```

Upon successful execution, the added trust policy is printed out. For example:

In text format

```text
name: "wabbit-networks-dev"
registryScopes: ["*"]
level: "strict"
trustStores: ["ca:wabbit-networks-dev"]
trustedIdentities: ["*"]
```

In json format

```json
{
  "name": "wabbit-networks-dev",                              
  "registryScopes": [ "*" ],                                  
  "signatureVerification": {                                  
      "level": "strict"
  },
  "trustStores": [ "ca:wabbit-networks-dev" ],                
  "trustedIdentities": [                                      
      "*"
  ]

> }

Reviewer 3: IMO we don't need to display both json and text format for each variation of a command, its making doc large and doesnt adds much value.

```

Users can also use `notation policy list` to confirm the trust policies are added.

### Add trust policies from a JSON file

> Users can add trust policies from a JSON file. This is specially useful when add multiple trust polices.

Reviewer 3: ```suggestion
Users can add trust policies from a JSON file. This is specially useful when user wants add multiple trust polices.
```

Create a JSON file that includes the trust polices. For example, a file named `mytrustpolicy.json` with the following content:

> ```json

Reviewer 3: ```suggestion
```jsonc
```

{
    "version": "1.0",
    "trustPolicies": [     
        // First policy                                             

>         {
>             "name": "wabbit-networks-dev",                              
>             "registryScopes": [ "dev.wabbitnetworks.io/net-monitor" ],  
>             "signatureVerification": {                                  
>                 "level": "strict"
>                 "override" : {
>                      "expiry" : "log",
>                      "authenticity": "log"
>                 }
>             },
>             "trustStores": [ "ca:wabbit-networks-dev" ],
>             "trustedIdentities": [
>                 "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev, CN=wabbit-networks.io"
>             ]
>         },

Reviewer 3: ```suggestion
        {
            "name": "wabbit-networks-dev",                              
            "registryScopes": [ "dev.wabbitnetworks.io/net-monitor" ],  
            "signatureVerification": {                                  
                "level": "strict"
            },
            "trustStores": [ "ca:wabbit-networks-dev" ],
            "trustedIdentities": [ "*" ]
        },
```

        // Second policy
        {},
        // Third policy

>         {}

Reviewer 3: THis is invalid trust-policy statement.

    ]
}
```

Then run the following command to add trust policies

```shell
notation policy add --file mytrustpolicy.json
```

The execution fails in one of below cases:

> - Version doesn't support. Currently only `1.0` version is supported.

Reviewer 3: ```suggestion
- Unsupported Version. Currently only `1.0` version is supported.
```

> - The trust policy name exists.

Reviewer 3: ```suggestion
- A trust policy with same name already exists.
```

  "registryScopes": [ "dev-2.wabbitnetworks.io/net-monitor" ],                                  
  "signatureVerification": {                                  
      "level": "permissive"
  },
  "trustStores": [ "ca:wabbit-networks-dev-2" ],                
  "trustedIdentities": [                                      
      "x509.subject: C=US, ST=WA, L=Seattle, O=Example, OU=Dev-2, CN=wabbit-networks.io"
  ]
}
```

### Update trust policies from a JSON file

Users can update trust policies from a JSON file. This is specially useful when update multiple trust polices.

Create a JSON file that includes the trust polices with updated values. Then run the following command to update trust policies.

```shell
notation policy update --file mytrustpolicy.json
```

The execution fails in one of below cases:

- Version doesn't support. Currently only `1.0` version is supported.
- The trust policy name doesn't exists.
- More than one trust policy that uses a global scope, that is, the value of `registryScopes` is `["*"]`.
- The values of `trustIdentities` overlap. For example, the following two identity values are overlapping:
  - "C=US, ST=WA, O=wabbit-network.io, OU=org1"

>   - "C=US, ST=WA, O=wabbit-network.io"

Reviewer 3: IMO this is redundant information and can be removed.

Upon successful execution, the updated trust policies are printed out. Users can also use `notation policy list` to confirm the trust policies are added.

### List all the trust policies by names

The output is a list of trust policy names.

```shell
notation policy list
```

An example of output messages:

in text format:

```text
name: wabbit-network-dev
registryScopes: ["dev.wabbitnetworks.io/net-monitor"]

name: wabbit-network-prod

> registryScopes: ["prod.wabbitnetworks.io/net-monitor"]

Reviewer 3: Should we use tree based representation here?
